include/irr/video/IGPUVirtualTexture.h

```diff
@@ -0,0 +1,188 @@
+            false
+            ),
+        m_driver(_driver)
+    {
+        //now copy from CPU counterpart resources that can be shared (i.e. just copy state) between CPU and GPU
+        //and convert to GPU those which can't be "shared": page table and VT resident storages along with their images and views
+
+        m_pageTable = std::move(_driver->getGPUObjectsFromAssets(&_cpuvt->getPageTable(), &_cpuvt->getPageTable()+1)->front());
+
+        m_freePageTableLayerIDs = _cpuvt->getFreePageTableLayersStack();
+
+        memcpy(m_layerToViewIndexMapping->data(), _cpuvt->getLayerToViewIndexMapping().begin(), 
+            _cpuvt->getLayerToViewIndexMapping().size()*sizeof(decltype(m_layerToViewIndexMapping)::pointee::value_type));
+
+        const auto& cpuStorages = _cpuvt->getResidentStorages();
+        for (const auto& pair : cpuStorages)
+        {
+            auto* cpuStorage = static_cast<asset::ICPUVirtualTexture::ICPUVTResidentStorage*>(pair.second.get());
+            const asset::E_FORMAT_CLASS fmtClass = pair.first;
+
+            m_storage.insert({fmtClass, core::make_smart_refctd_ptr<IGPUVTResidentStorage>(_driver, _am, cpuStorage)});
+        }
+
+        m_fsamplers.views = _cpuvt->getFloatViews().size() ? _driver->getGPUObjectsFromAssets(_cpuvt->getFloatViews().begin(), _cpuvt->getFloatViews().end()) : nullptr;
+        m_isamplers.views = _cpuvt->getIntViews().size() ? _driver->getGPUObjectsFromAssets(_cpuvt->getIntViews().begin(), _cpuvt->getIntViews().end()) : nullptr;
+        m_usamplers.views = _cpuvt->getUintViews().size() ? _driver->getGPUObjectsFromAssets(_cpuvt->getUintViews().begin(), _cpuvt->getUintViews().end()) : nullptr;
+    }
+
+    STextureData pack(const image_t* _img, const asset::IImage::SSubresourceRange& _subres, asset::ISampler::E_TEXTURE_CLAMP _wrapu, asset::ISampler::E_TEXTURE_CLAMP _wrapv, asset::ISampler::E_TEXTURE_BORDER_COLOR _borderColor) override
+    {
+        assert(0);
+        return STextureData::invalid();
+    }
+
+    bool free(const STextureData& _addr, const asset::IImage* _img, const asset::IImage::SSubresourceRange& _subres) override
+    {
+        assert(0);
+        return false;
+    }
+
+    core::smart_refctd_ptr<IGPUImageView> createPageTableView() const override
+    {
+        IGPUImageView::SCreationParams params;
+        params.flags = static_cast<IGPUImageView::E_CREATE_FLAGS>(0);
+        params.format = m_pageTable->getCreationParameters().format;
+        params.subresourceRange.aspectMask = static_cast<asset::IImage::E_ASPECT_FLAGS>(0);
+        params.subresourceRange.baseArrayLayer = 0u;
+        params.subresourceRange.layerCount = m_pageTable->getCreationParameters().arrayLayers;
+        params.subresourceRange.baseMipLevel = 0u;
+        params.subresourceRange.levelCount = m_pageTable->getCreationParameters().mipLevels;
+        params.viewType = asset::IImageView<IGPUImage>::ET_2D_ARRAY;
+        params.image = m_pageTable;
```

Comment:
this part is common between ICPU and IGPU classes, only depends on the template parameter, factor it out.